lib/public/HintException.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
+ *
+ * @author Arthur Schiwon <blizzz@arthur-schiwon.de>
+ * @author Bart Visscher <bartv@thisnet.nl>
+ * @author Lukas Reschke <lukas@statuscode.ch>
+ * @author Michael Gapczynski <GapczynskiM@gmail.com>
+ * @author Morris Jobke <hey@morrisjobke.de>
+ * @author Thomas Müller <thomas.mueller@tmit.eu>
+ *
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP;
+
+/**
+ * Class HintException
+ *
+ * An Exception class with the intention to be presented to the end user
+ *
+ * @package OCP
+ * @since 22.0.0
+ */
+class HintException extends \Exception {
+	private $hint;
+
+	/**
+	 * HintException constructor.
+	 *
+	 * @param string $message  The error message. It will be not revealed to the
+	 *                         the user (unless the hint is empty) and thus
+	 *                         should be not translated.
+	 * @param string $hint     A useful message that is presented to the end
+	 *                         user. It should be translated, but must not
+	 *                         contain sensitive data.
+	 * @param int $code
+	 * @param \Exception|null $previous
+	 * @since 22.0.0
+	 */
+	public function __construct($message, $hint = '', $code = 0, \Exception $previous = null) {
```

Comment:
```suggestion
	public function __construct(string $message, string $hint = '', int $code = 0, ?\Exception $previous = null) {
```
Can we do this or will it do boom?

Reply:
Will do boom because of the exceptions that extend it and then they don't match anymore 🙈